### DISCUSSION TITLE
ConnectTimeout does not cancel upgrade of TcpConnection to SslConnection

### DISCUSSION THREAD

**Discussion by joao-genebra** [nats-io/nats.net#335] [General]
When opening a connection, after the connection has been created and before it is upgraded to ssl [here](https://github.com/nats-io/nats.net.v2/blob/3e3c86745008cf190e6f16d83a08a6a839adfda4/src/NATS.Client.Core/NatsConnection.cs#L288C21-L294C22), if some kind of hiccup occurs in our network, the timeout value set in `ConnectTimeout` of `NatsOpts` is not stopping the connection upgrade, which is causing the call to `ConnectAsync` to stay blocked.

We know that the connection has already been opened because we can find [this](https://github.com/nats-io/nats.net.v2/blob/3e3c86745008cf190e6f16d83a08a6a839adfda4/src/NATS.Client.Core/Internal/NatsReadProtocolProcessor.cs#L392) log.

**Comment by mtmk** [comment]
I'm guessing the call to `AuthenticateAsClientAsync` is blocking. I wonder if passing `CancellationToken` with timeout would help.

https://github.com/nats-io/nats.net.v2/blob/f326e746f570bd45ac3641cd970c7e2539e51b8b/src/NATS.Client.Core/Internal/SslStreamConnection.cs#L88

**Comment by mtmk** [comment]
@joao-genebra which version of the client are you using btw. We have some big changes in 2.1.0-preview.1

**Comment by joao-genebra** [comment]
The NATS client version we are using is 2.0.1. We would rather not use a preview version and prefer to wait for a stable version.

Also, the behavior seems to be flaky and is only happening in production, so we are not able to replicate it locally.

Regarding your suggestion of passing a `CancellationToken`, it would be better if we were able to pass it directly to `ConnectAsync`. 

Here are the logs showing what I had described earlier:
![chrome_mIbS5MJ1Zq](https://github.com/nats-io/nats.net.v2/assets/84010896/0a4f4a9e-ea21-42a7-80cf-099e5ef9ad78)

**Comment by mtmk** [comment]
Thanks for the report @joao-genebra. it looks like a bug.

Opened an issue for this #337 please follow up there.

**Comment by mtmk** [comment]
@joao-genebra sorry the fix delayed a little because of an unrelated issue. fix to this issue is in [2.1.0-preview.4](https://github.com/nats-io/nats.net.v2/releases/tag/v2.1.0-preview.4). it'd be great if you have the opportunity to test it. Thank you!
